cpp/include/cudf/detail/timezone.cuh

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/timezone.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+
+namespace cudf::detail {
+
+/**
+ * @brief Returns the GMT offset for a given date and given timezone table.
+ *
+ * @param ttimes Transition times; trailing `cycle_entry_cnt` entries are used for all times
+ * beyond the one covered by the TZif file
+ * @param offsets Time offsets in specific intervals; trailing `cycle_entry_cnt` entries are used
+ * for all times beyond the one covered by the TZif file
+ * @param ts ORC timestamp
+ *
+ * @return GMT offset
+ */
+inline __device__ duration_s get_gmt_offset(table_device_view tz_table, timestamp_s ts)
+{
+  if (tz_table.num_rows() == 0) { return duration_s{0}; }
+
+  cudf::device_span<timestamp_s const> ttimes(tz_table.column(0).head<timestamp_s>(),
```

Comment:
Can we give this a longer name? Does `ttimes` mean `transition_times`? Let's try to increase clarity as much as possible in names.